File: src/tros.erl
%%% @copyright 2016+ Hippware, Inc.
%%% @doc TROS helper functions
-module(tros).

-export([parse_url/1,
         make_jid/2,
         make_url/2,

         get_metadata/2,
         get_access/1,
         get_owner/1,

         delete/2,
         keep/2
        ]).

-type file_id() :: ejabberd:lresource().
-type url() :: <<_:40,_:_*8>>.
-type metadata() :: mod_wocky_tros_backend:metadata().
-type result(ResultType) :: {ok, ResultType} | {error, atom()}.

-export_type([file_id/0, url/0, metadata/0, result/1]).

-include_lib("ejabberd/include/jlib.hrl").
-include_lib("ejabberd/include/ejabberd.hrl").

-spec parse_url(url()) -> {ok, {ejabberd:lserver(), ejabberd:lresource()}} |
                          {error, invalid_url}.
parse_url(<<"tros:", JID/binary>>) ->
    #jid{lserver = Server, lresource = Resource} = jid:from_binary(JID),
    case Resource of
        <<"file/", FileID/binary>> -> {ok, {Server, FileID}};
        _ -> {error, invalid_url}
    end;
parse_url(_) ->
    {error, invalid_url}.

-spec make_jid(ejabberd:lserver(), ejabberd:resource()) -> ejabberd:jid().
make_jid(Server, FileID) ->
    jid:make(<<>>, Server, <<"file/", FileID/binary>>).

-spec make_url(ejabberd:lserver(), file_id()) -> url().
make_url(Server, FileID) ->
    <<"tros:", (jid:to_binary(make_jid(Server, FileID)))/binary>>.

-spec get_owner(metadata()) -> result(ejabberd:luser()).
get_owner(Metadata) ->
    get_file_info(get_owner, [Metadata]).

-spec get_access(metadata()) -> result(binary()).
get_access(Metadata) ->
    get_file_info(get_access, [Metadata]).

-spec get_metadata(ejabberd:lserver(), file_id()) -> result(metadata()).
get_metadata(Server, FileID) ->
    get_file_info(get_metadata, [Server, FileID]).

get_file_info(Function, Args) ->
    apply(mod_wocky_tros:backend(), Function, Args).

-spec keep(ejabberd:lserver(), file_id()) -> ok | {error, not_found}.
keep(Server, FileID) ->
    apply(mod_wocky_tros:backend(), keep, [Server, FileID]).

-spec delete(ejabberd:lserver(), file_id()) -> ok.
delete(Server, FileID) ->
    apply(mod_wocky_tros:backend(), delete, [Server, FileID]).

File: apps/wocky/lib/wocky/tros.ex
# @copyright 2016+ Hippware, Inc.
defmodule Wocky.TROS do
  @moduledoc "API module for the TROS file handling system."

  use Wocky.JID

  alias Wocky.Repo.ID
  alias Wocky.TROS.Metadata

  @type owner     :: JID.luser
  @type server    :: JID.lserver
  @type file_id   :: JID.lresource
  @type url       :: binary # <<_:40,_:_*8>>
  @type metadata  :: map # %{binary => binary}
  @type file_type :: :full | :original | :thumbnail

  @type error :: :not_found | :metadata_not_found | {:retrieve_error, binary}
  @type result(type) :: {:ok, type} | {:error, error}

  @callback delete(server, file_id) :: :ok
  @callback make_upload_response(JID.t, file_id, integer, metadata) ::
    {list, list}
  @callback make_download_response(server, file_id) :: {list, list}
  @callback get_download_url(server, file_id) :: url

  @thumbnail_suffix "-thumbnail"
  @original_suffix "-original"

  @spec make_jid(server, file_id) :: JID.t
  def make_jid(server, file_id),
    do: make_jid("", server, file_id)

  @spec make_jid(owner, server, file_id) :: JID.t
  def make_jid(owner, server, file_id),
    do: JID.make(owner, server, "file/#{file_id}")

  @spec parse_url(url) :: {:ok, {server, file_id}} | {:error, :invalid_url}
  def parse_url("tros:" <> jid) do
    jid(lserver: server, lresource: resource) = JID.from_binary(jid)
    case resource do
      "file/" <> file_id -> {:ok, {server, file_id}}
      _ -> {:error, :invalid_url}
    end
  end
  def parse_url(_), do: {:error, :invalid_url}

  @spec make_url(server, file_id) :: url
  def make_url(server, file_id),
    do: server |> make_jid(file_id) |> make_url

  @spec make_url(JID.t) :: url
  def make_url(jid),
    do: "tros:#{JID.to_binary(jid)}"

  @spec get_base_id(file_id) :: file_id
  def get_base_id(file_id) do
    file_id
    |> String.replace_suffix(@thumbnail_suffix, "")
    |> String.replace_suffix(@original_suffix, "")
  end

  @spec get_type(file_id) :: file_type
  def get_type(file_id) do
    if String.ends_with?(file_id, @original_suffix) do
      :original
    else
      if String.ends_with?(file_id, @thumbnail_suffix) do
        :thumbnail
      else
        :full
      end
    end
  end

  @spec variants(file_id) :: [binary]
  def variants(file_id) do
    for suffix <- ["", @original_suffix, @thumbnail_suffix] do
      file_id <> suffix
    end
  end

  @spec get_metadata(file_id) :: result(Metadata.t)
  def get_metadata(id) do
    if ID.valid?(id) do
      case Metadata.get(id) do
        nil -> {:error, :not_found}
        value -> {:ok, value}
      end
    else
      {:error, :not_found}
    end
  end

  @spec update_access(file_id, binary) :: result(Metadata.t)
  def update_access(file_id, new_access) do
    Metadata.set_access(file_id, new_access)
  end

  @spec delete(server, file_id) :: :ok
  def delete(server, file_id) do
    Metadata.delete(file_id)
    (backend()).delete(server, file_id)
  end

  @spec make_upload_response(JID.t, file_id, integer, binary, metadata) ::
    {list, list}
  def make_upload_response(owner_jid, file_id, size, access, meta) do
    jid(luser: owner) = owner_jid
    {:ok, _} = Metadata.put(file_id, owner, access)

    (backend()).make_upload_response(owner_jid, file_id, size, meta)
  end

  @spec make_download_response(server, file_id) :: {list, list}
  def make_download_response(server, file_id),
    do: (backend()).make_download_response(server, file_id)

  @spec ready?(file_id) :: boolean
  def ready?(file_id), do: Metadata.ready?(file_id)

  @spec get_download_url(server, file_id) :: binary
  def get_download_url(server, file_id) do
    (backend()).get_download_url(server, file_id)
  end

  def thumbnail_id(file_id), do: file_id <> @thumbnail_suffix

  defp backend do
    Application.fetch_env!(:wocky, :tros_backend)
  end
end
